File: its-my-training-day/leetcode-top-75/questions/trees/tree-bst/validate-bst/solution.py
import os
import sys
from typing import List, Dict, Optional, Tuple

ROOT_DIR = os.path.abspath(os.path.join(os.path.dirname(__file__), "../../../../"))
if ROOT_DIR not in sys.path:
    sys.path.append(ROOT_DIR)

from shared.python.ds import ArrayList, ListNode, Stack, Queue, TreeNode, MinHeap, MaxHeap

def validate_bst(root: Optional[TreeNode]) -> bool:
    def helper(node: Optional[TreeNode], low: int, high: int) -> bool:
        if node is None:
            return True
        if node.val <= low or node.val >= high:
            return False
        return helper(node.left, low, node.val) and helper(node.right, node.val, high)

    return helper(root, -10**18, 10**18)

File: its-my-training-day/leetcode-top-75/questions/trees/tree-bst/validate-bst/solution.cpp
#include "../../../../shared/cpp/ds.hpp"

#include <algorithm>
#include <cctype>
#include <cstdint>
#include <functional>
#include <string>
#include <unordered_map>
#include <vector>

bool validateBST(TreeNode* root) {
    std::function<bool(TreeNode*, long long, long long)> helper = [&](TreeNode* node, long long low, long long high) {
        if (!node) return true;
        if (node->val <= low || node->val >= high) return false;
        return helper(node->left, low, node->val) && helper(node->right, node->val, high);
    };
    return helper(root, -1000000000000LL, 1000000000000LL);
}
